### PR TITLE
feature: Support lua_ssl_certificate and lua_ssl_certificate_key

### PR DESCRIPTION
**Pull request by ZJfans**
Currently, this module only supports one-way verification of upstream certificates. This PR is to support using sll two-way authentication to connect upstream, such as redis.   I added 2 keywords, **lua_ssl_certificate** and **lua_ssl_certificate_key**，using redis as an example (redis need to enable ssl two-way authentication):

```
server{
    listen 80;

lua_ssl_certificate /openresty/nginx/ssl/test.crt;
lua_ssl_certificate_key /openresty/nginx/ssl/test.key;

lua_ssl_trusted_certificate /openresty/nginx/ssl/ca.crt;

location /redis {
        content_by_lua_block {
            local redis = require "resty.redis"
            local red = redis:new()

            local ok, err = red:connect("127.0.0.1", 6379,  {
                ssl = true,
                ssl_verify = true
            })
            if not ok then
                ngx.say("failed to connect: ", err)
                return
            end

            ngx.say("ok")
        }
    }
}
```

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

**Comment by zhuizhuhaomeng** [comment]
@ZJfans  need to add test cases for the new directive.

**Comment by ZJfans** [comment]
> @ZJfans need to add test cases for the new directive.

I added test case 187

**Comment by zhuizhuhaomeng** [comment]
@ZJfans we have merged the PR 'commit 592a91c1ab4079' which conflicts with your PR.
Please rebase your PR to the latest master.

**Comment by zhuizhuhaomeng** [comment]
Please go to https://github.com/openresty/lua-nginx-module/pull/2196 for more details.
And you may need to do the same optimization like that PR.

**Comment by zhuizhuhaomeng** [comment]
https://app.travis-ci.com/github/openresty/lua-nginx-module/jobs/604417033#L2346
The build failed @ZJfans

**Comment by ZJfans** [comment]
> https://app.travis-ci.com/github/openresty/lua-nginx-module/jobs/604417033#L2346 The build failed @ZJfans

A little strange,I need to research this error,do you have any suggestions

**Comment by ZJfans** [comment]
> > https://app.travis-ci.com/github/openresty/lua-nginx-module/jobs/604417033#L2346 The build failed @ZJfans
> 
> A little strange,I need to research this error,do you have any suggestions

Now my test case is wrong,i need a few days to fix it

File: t/187-ssl-two-verification.t [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+run_tests();
+
+__DATA__
+
+=== TEST 1: ssl connection (without certificate and certificate_key)
+--- config
+    location /t {
+        content_by_lua_block {
+            local redis = require "resty.redis"
```
[the remark this answers is not in the document]
Comment (ZJfans) [reply t/187-ssl-two-verification.t]: > We don't want to import too much library in lua-nginx-module. You can use tcp socket directly. Go to https://github.com/openresty/lua-nginx-module#ngxsockettcp for more detail.

bcc29e64b4844cbbbe80ba685515d0a
The test fails without any information,should I check my code or test case

File: t/187-ssl-two-verification.t [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+$Test::Nginx::Util::NginxVersion >= 1.017009 ?
+qr/reusable connection: 0
+ssl client hello: connection reusable: 0
+ssl_client_hello_by_lua\(nginx.conf:\d+\):1: ssl client hello by lua is running!,/
+: qr /reusable connection: 1
+ssl client hello: connection reusable: 1
+reusable connection: 0
+ssl_client_hello_by_lua\(nginx.conf:\d+\):1: ssl client hello by lua is running!,/
+
+
```
Comment (zhuizhuhaomeng) [inline t/187-ssl-two-verification.t:136]: extra blank line

File: t/187-ssl-two-verification.t [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+repeat_each(3);
+
+# All these tests need to have new openssl
+my $NginxBinary = $ENV{'TEST_NGINX_BINARY'} || 'nginx';
+my $openssl_version = eval { `$NginxBinary -V 2>&1` };
+
+if ($openssl_version =~ m/built with OpenSSL (0\S*|1\.0\S*|1\.1\.0\S*)/) {
+    plan(skip_all => "too old OpenSSL, need 1.1.1, was $1");
+} else {
+    plan tests => repeat_each() * (blocks() * 6 + 6);
```
Comment (zhuizhuhaomeng) [inline t/187-ssl-two-verification.t:14]:
```suggestion
    plan tests => repeat_each() * (blocks() * 6);
```

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

**Comment by zhuizhuhaomeng** [comment]
@ZJfans I added two comments for the PR.
And please add doc for the new directives.

**Comment by ZJfans** [comment]
> @ZJfans I added two comments for the PR. And please add doc for the new directives.

Thank you for solving my previous problem , and i will add doc for this directives.

**Comment by ZJfans** [comment]
> @ZJfans I added two comments for the PR. And please add doc for the new directives.

I add doc for the two directives, and what else needs to be done now, do i need add more test cases?

**Comment by zhuizhuhaomeng** [comment]
@ZJfans I made some modifications to the test case.
ssl_certificate&ssl_certificate_key are directives for the virtual server.  
So I think lua_ssl_certificte & lua_ssl_certificate_key can be confused.
But I can not figure out a better name for now.

**Comment by ZJfans** [comment]
@zhuizhuhaomeng New users will be confused, it`s like ssl_trusted_certificate & lua_ssl_trusted_certificate? i think these directives should keep the style of openresty, and i considered lua_upstream_ssl_certificate, but i don`t think it`s such as a good idea

**Comment by zhuizhuhaomeng** [comment]
That makes sense.

File: doc/HttpLuaModule.wiki [hunk 4]
```diff
@@ -2845,6 +2845,38 @@ The support for the <code>TLSv1.3</code> parameter requires version <code>v0.10.
...
+
+'''default:''' ''no''
+
+'''context:''' ''http, server, location''
+
+Specifies the file path to the private key associated with the SSL/TLS certificate used in the [[#tcpsock:sslhandshake|tcpsock:sslhandshake]] method.
+
+This directive allows you to specify the private key file corresponding to the SSL/TLS certificate specified by lua_ssl_certificate. The private key should be in PEM format and must match the certificate.
+
+This directive was first introduced in the <code>v0.10.25</code> release.
```
Comment (zhuizhuhaomeng) [inline doc/HttpLuaModule.wiki:2876]:
```suggestion
This directive was first introduced in the <code>v0.10.26</code> release.
```

File: doc/HttpLuaModule.wiki [hunk 5]
```diff
@@ -2845,6 +2845,38 @@ The support for the <code>TLSv1.3</code> parameter requires version <code>v0.10.
...
+
+'''default:''' ''no''
+
+'''context:''' ''http, server, location''
+
+Specifies the file path to the SSL/TLS certificate in PEM format used for the [[#tcpsock:sslhandshake|tcpsock:sslhandshake]] method.
+
+This directive allows you to specify the SSL/TLS certificate that will be presented to server during the SSL/TLS handshake process.
+
+This directive was first introduced in the <code>v0.10.25</code> release.
```
Comment (zhuizhuhaomeng) [inline doc/HttpLuaModule.wiki:2860]:
```suggestion
This directive was first introduced in the <code>v0.10.26</code> release.
```

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

File: doc/HttpLuaModule.wiki (hunk 5 above)
Comment (ZJfans) [reply doc/HttpLuaModule.wiki]: ok，i update the version in the doc ，and should i merge all my branches?

Comment (ZJfans) [reply doc/HttpLuaModule.wiki]: > 

@zhuizhuhaomeng

**Review by zhuizhuhaomeng — APPROVED** [review]

**Pull request merged**